### PR TITLE
add an example JWT auth verification plugin

### PR DESCRIPTION
**Pull request by garypen**
Implemented support for 3 common HMAC algorithms.
    
There's potentially a lot of extra things that could be added to this
plugin:
 - Additional algorithms
 - Custom Claims via configuration
 - JWK validation
 - Token refresh
 - ...
    
I think that is probably material for a different issue...

fixes: #434

### CODE REVIEW

**Comment by garypen** [comment]
> @garypen your pull request is missing a changelog!

It is. That's because it's an example and they don't appear in the changelog (or at least not AFAICT)

**Review by BrynCooke — APPROVED** [review]
Looks great!

File: examples/jwt-auth/src/jwt.rs [hunk 1]
```diff
@@ -0,0 +1,769 @@
...
+                        .build()
+                        .with_status(status);
+                    Ok(ControlFlow::Break(res))
+                }
+
+                // The http_request is stored in a `RouterRequest` context.
+                // We are going to check the headers for the presence of the header we're looking for
+                // We are implementing: https://www.rfc-editor.org/rfc/rfc6750
+                // so check for our AUTHORIZATION header.
+                if !req.context.request.headers().contains_key(AUTHORIZATION) {
```
Comment (Geal) [inline examples/jwt-auth/src/jwt.rs:256]: why not do this check and the next in one? e.g.

```rust
let jwt_value_result = match req
                    .context
                    .request
                    .headers()
                    .get(AUTHORIZATION) {
    None => 
    // Prepare an HTTP 401 response with a GraphQL error message
    return failure_message(
        format!("Missing '{}' header", AUTHORIZATION),
        StatusCode::UNAUTHORIZED,
    ),
    Some(value) => value.to_str(),
};
```

**Review by Geal — APPROVED** [review]

File: examples/jwt-auth/src/jwt.rs (hunk 1 above)
Comment (garypen) [reply examples/jwt-auth/src/jwt.rs]: Done: 235fd34435a22b3df5d74473c50fb3be5dd96ffa

**Review by Geal — APPROVED** [review]

File: examples/jwt-auth/config.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  # max_token_life: is the time between issued and expires for a token
+  #
+  # For example: we want to use HS256 as our verification algorithm and we are
+  # prepared to let tokens have an extra 60 seconds of life, but we won't
+  # accept tokens that were issued with more than 10 minutes of life.
+  example.jwt:
+    algorithm: HS256
+    key: 629709bdc3bd794312ccc3a1c47beb03ac7310bc02d32d4587e59b5ad81c99ba
+    time_tolerance: 60
+    max_token_life: 600
```
Comment (o0Ignition0o) [inline examples/jwt-auth/config.yml:29]: Really like the documentation section, it shows a lot of care 🤗

File: examples/jwt-auth/src/jwt.rs [hunk 3]
```diff
@@ -0,0 +1,758 @@
+//! Example implementation of a JWT verifying plugin
+//!
+//! DISCLAIMER:
+//!     This is an example for illustrative purposes. It has not been secrity audited
```
Comment (o0Ignition0o) [inline examples/jwt-auth/src/jwt.rs:4]:
```suggestion
//!     This is an example for illustrative purposes. It has not been security audited
```

File: examples/jwt-auth/src/jwt.rs [hunk 4]
```diff
@@ -0,0 +1,758 @@
...
+                            // Prepare an HTTP 403 response with a GraphQL error message
+                            return failure_message(
+                                format!("{jwt} is not authorized: {}", err),
+                                StatusCode::FORBIDDEN,
+                            );
+                        }
+                    }
+                } else {
+                    // We've only implemented support for HMAC. Two possibilities here:
+                    //  - Either we tried to specify a valid, but unimplemented algorithem
```
Comment (o0Ignition0o) [inline examples/jwt-auth/src/jwt.rs:364]:
```suggestion
                    //  - Either we tried to specify a valid, but unimplemented algorithm
```

File: examples/jwt-auth/src/jwt.rs [hunk 5]
```diff
@@ -0,0 +1,758 @@
...
+                    //  - Or we have specified an invalid algorithem (typo?). Either way:
```
Comment (o0Ignition0o) [inline examples/jwt-auth/src/jwt.rs:365]:
```suggestion
                    //  - Or we have specified an invalid algorithm (typo?). Either way:
```

**Review by o0Ignition0o — APPROVED** [review]
Love it! Great job!

File: examples/jwt-auth/src/jwt.rs (hunk 5 above)
Comment (garypen) [reply examples/jwt-auth/src/jwt.rs]: Done: 89b79951428f8bb1a1c09e3e89924b763637aef6

File: examples/jwt-auth/src/jwt.rs (hunk 4 above)
Comment (garypen) [reply examples/jwt-auth/src/jwt.rs]: Done: 89b79951428f8bb1a1c09e3e89924b763637aef6

File: examples/jwt-auth/src/jwt.rs (hunk 3 above)
Comment (garypen) [reply examples/jwt-auth/src/jwt.rs]: Done: 89b79951428f8bb1a1c09e3e89924b763637aef6

**Pull request merged**